File: matlab/qr/main_Ainnerprod.m
%
   close all
   clear
%
   oldpath = path;
   path('../orth/',oldpath)  
   oldpath = path;
   path('../matrix_market/',oldpath)  
%
   m = 100;
   n = 50;
%
   M = rand(m);
   M = 0.5 * (M + M');
   M = M + (m * eye(m));
%
   range_log10KA = 0:1:16;
%
   condA               = zeros(size(range_log10KA));
%
   orth_mgs_lvl1       = zeros(size(range_log10KA));
   orth_mgs_lvl2       = zeros(size(range_log10KA));
   orth_mgs_lvl2_2sync = zeros(size(range_log10KA));
   orth_mgs_lvl2_opv1 = zeros(size(range_log10KA));
   orth_mgs_lvl2_opv2 = zeros(size(range_log10KA));
%
   repres_mgs_lvl1       = zeros(size(range_log10KA));
   repres_mgs_lvl2       = zeros(size(range_log10KA));
   repres_mgs_lvl2_2sync = zeros(size(range_log10KA));
   repres_mgs_lvl2_opv1 = zeros(size(range_log10KA));
   repres_mgs_lvl2_opv2 = zeros(size(range_log10KA));
%
   s_mgs_lvl1            = zeros(size(range_log10KA));
   s_mgs_lvl2            = zeros(size(range_log10KA));
   s_mgs_lvl2_2sync      = zeros(size(range_log10KA));
   s_mgs_lvl2_opv1      = zeros(size(range_log10KA));
   s_mgs_lvl2_opv2      = zeros(size(range_log10KA));
%
   i = 0;
   for log10KA = range_log10KA
      i = i+1;
%	   
      U = randn(m,n); [U,~]=qr(U,0);
      V = randn(n,n); [V,~]=qr(V,0);
      S = diag( 10.^( linspace( 0, log10KA, n ) ) );
      A = U * S * V';
      clear U S V;
      nrmA = norm(A,'fro');
      condA(i) = cond(A);
%
     [ MQ, Q, R ] = qr__orth_mgs_lvl1_Ainnerprod( A, M ); [ repres_mgs_lvl1(i), orth_mgs_lvl1(i), s_mgs_lvl1(i) ] = qr_Ainnerprod_orth_stabilitymetric( A, M, Q, R ); clear MQ Q R;
     [ MQ, Q, R ] = qr__orth_mgs_lvl2_Ainnerprod( A, M ); [ repres_mgs_lvl2(i), orth_mgs_lvl2(i), s_mgs_lvl2(i) ] = qr_Ainnerprod_orth_stabilitymetric( A, M, Q, R ); clear MQ Q R;
     [ MQ, Q, R ] = qr__orth_mgs_lvl2_Ainnerprod_2sync( A, M ); [ repres_mgs_lvl2_2sync(i), orth_mgs_lvl2_2sync(i), s_mgs_lvl2_2sync(i) ] = qr_Ainnerprod_orth_stabilitymetric( A, M, Q, R ); clear MQ Q R;
     [ MQ, Q, R ] = qr__orth_mgs_lvl2_Ainnerprod_orth_projector( A, M ); [ repres_mgs_lvl2_opv1(i), orth_mgs_lvl2_opv1(i), s_mgs_lvl2_opv1(i) ] = qr_Ainnerprod_orth_stabilitymetric( A, M, Q, R ); clear MQ Q R;
     [ MQ, Q, R ] = qr__orth_mgs_lvl2_Ainnerprod_orth_projector_v2( A, M ); [ repres_mgs_lvl2_opv2(i), orth_mgs_lvl2_opv2(i), s_mgs_lvl2_opv2(i) ] = qr_Ainnerprod_orth_stabilitymetric( A, M, Q, R ); clear MQ Q R;
%
   end
%
   figure
   ax = axes;
   x = 10.^(range_log10KA);
   loglog(x,condA.^2*eps,'--','LineWidth', 2, 'color', 'k','HandleVisibility', 'off'); hold on;
   loglog(x,condA*eps,'--','LineWidth', 2, 'color', 'k','HandleVisibility', 'off'); hold on;
   loglog(x,orth_mgs_lvl1,'-s','MarkerSize', 12, 'Color', '#f5e284', 'LineWidth', 2 ); hold on;
   loglog(x,orth_mgs_lvl2,'-s','MarkerSize', 12, 'Color', '#4D0F6D', 'LineWidth', 2 ); hold on;
   loglog(x,orth_mgs_lvl2_2sync,'-s','MarkerSize', 12, 'Color', '#0F6D19', 'LineWidth', 2 ); hold on;
   loglog(x,orth_mgs_lvl2_opv1,'-s','MarkerSize', 12, 'Color', '#D0761B', 'LineWidth', 2 ); hold on;
   loglog(x,orth_mgs_lvl2_opv2,'-s','MarkerSize', 12, 'Color', '#996633', 'LineWidth', 2 ); hold on;
   grid on;
   axis([ min(x) max(x) 1e-16 1e2 ]);
   legend('MGS lvl1','MGS lvl2','MGS lvl2 2-sync','MGS lvl2 Orthogonal Projection V1','MGS lvl2 Orthogonal Projection V2','Location','southeast');
   title({'Orthogonality Comparison, QR factorization','m = 500, n = 100'}, 'FontWeight', 'bold');
   xlabel('Condition Number of Input Matrix', 'FontWeight', 'bold')
   ylabel('Level of Orthogonality, || I - Q^T Q ||_f_r_o', 'FontWeight', 'bold')
   set(gca,'FontSize',12);   
   set(gca,'FontWeight','bold');   
%
return
   figure
   ax = axes;
   x = 10.^(range_log10KA);
   loglog(x,condA.^2*eps,'--','LineWidth', 2, 'color', 'k','HandleVisibility', 'off'); hold on;
   loglog(x,condA*eps,'--','LineWidth', 2, 'color', 'k','HandleVisibility', 'off'); hold on;
   loglog(x,repres_mgs_lvl1,'-s','MarkerSize', 12, 'Color', '#f5e284', 'LineWidth', 2 ); hold on;
   loglog(x,repres_mgs_lvl2,'-s','MarkerSize', 12, 'Color', '#4D0F6D', 'LineWidth', 2 ); hold on;
   loglog(x,repres_mgs_lvl2_2sync,'-s','MarkerSize', 12, 'Color', '#0F6D19', 'LineWidth', 2 ); hold on;
   grid on;
   axis([ min(x) max(x) 1e-17 1e2 ]);
   legend('MGS lvl1','MGS lvl2','MGS lvl2 2-sync','Location','east');
   title({'Representativity Comparison, QR factorization','m = 500, n = 100'}, 'FontWeight', 'bold');
   xlabel('Condition Number of Input Matrix', 'FontWeight', 'bold')
   ylabel('Level of Representativity, || A - Q R ||_f_r_o / || A ||_f_r_o', 'FontWeight', 'bold')
   set(gca,'FontSize',12);   
   set(gca,'FontWeight','bold');   
%

File: julia/qr/main_Ainnerprod.jl
include("qr__orth_mgs_lvl1_Ainnerprod.jl")
include("qr__orth_mgs_lvl2_Ainnerprod.jl")
include("qr__orth_mgs_lvl2_Ainnerprod_2sync.jl")
include("qr__orth_mgs_lvl2_Ainnerprod_orth_projector.jl")
include("qr__orth_mgs_lvl2_Ainnerprod_orth_projector_v2.jl")
include("qr_Ainnerprod_orth_stabilitymetric.jl")
using LinearAlgebra, Plots

m = 100
n = 50
M = rand(m, m)
M = 0.5 * (M + M')
M = M + m * Matrix(I, m, m)

range_log10KA = 0:1:16
condA              = zeros(length(range_log10KA))
orth_mgs1          = zeros(length(range_log10KA))
orth_mgs2          = zeros(length(range_log10KA))
orth_mgs2_2sync    = zeros(length(range_log10KA))
orth_mgs2_opv1     = zeros(length(range_log10KA))
orth_mgs2_opv2     = zeros(length(range_log10KA))
repres_mgs1        = zeros(length(range_log10KA))
repres_mgs2        = zeros(length(range_log10KA))
repres_mgs2_2sync  = zeros(length(range_log10KA))
repres_mgs2_opv1   = zeros(length(range_log10KA))
repres_mgs2_opv2   = zeros(length(range_log10KA))

for (i, log10KA) in enumerate(range_log10KA)
    U, _ = qr(randn(m, n))
    V, _ = qr(randn(n, n))
    S = Diagonal(10 .^ LinRange(0, log10KA, n))
    A = Matrix(U) * Matrix(S) * Matrix(V)'
    condA[i] = cond(A)

    MQ, Q, R = qr__orth_mgs_lvl1_Ainnerprod(A, M);                    repres_mgs1[i],       orth_mgs1[i],       _ = qr_Ainnerprod_orth_stabilitymetric(A, M, Q, R)
    MQ, Q, R = qr__orth_mgs_lvl2_Ainnerprod(A, M);                    repres_mgs2[i],       orth_mgs2[i],       _ = qr_Ainnerprod_orth_stabilitymetric(A, M, Q, R)
    MQ, Q, R = qr__orth_mgs_lvl2_Ainnerprod_2sync(A, M);              repres_mgs2_2sync[i], orth_mgs2_2sync[i], _ = qr_Ainnerprod_orth_stabilitymetric(A, M, Q, R)
    MQ, Q, R = qr__orth_mgs_lvl2_Ainnerprod_orth_projector(A, M);     repres_mgs2_opv1[i],  orth_mgs2_opv1[i],  _ = qr_Ainnerprod_orth_stabilitymetric(A, M, Q, R)
    MQ, Q, R = qr__orth_mgs_lvl2_Ainnerprod_orth_projector_v2(A, M);  repres_mgs2_opv2[i],  orth_mgs2_opv2[i],  _ = qr_Ainnerprod_orth_stabilitymetric(A, M, Q, R)
end

x = 10 .^ collect(range_log10KA)
p = plot(x, orth_mgs1,       xscale=:log10, yscale=:log10, label="MGS lvl1",       lw=2, marker=:utriangle)
plot!(p, x, orth_mgs2,       label="MGS lvl2",       lw=2, marker=:square)
plot!(p, x, orth_mgs2_2sync, label="MGS lvl2 2sync", lw=2, marker=:diamond)
plot!(p, x, orth_mgs2_opv1,  label="MGS lvl2 OP v1", lw=2, marker=:circle)
plot!(p, x, orth_mgs2_opv2,  label="MGS lvl2 OP v2", lw=2, marker=:star5)
xlabel!(p, "Condition Number of Input Matrix")
ylabel!(p, "M-Orthogonality || I - Q'MQ ||_fro")
title!(p, "A-Inner Product Orthogonality, QR factorization\nm = $m, n = $n")
display(p)
